core/src/main/java/fr/sncf/osrd/speedcontroller/generators/ConstructionAllowanceGenerator.java

```diff
@@ -56,59 +54,43 @@ protected double getFirstGuess() {
-                var updatesMap = getStepsAtPositionsToTarget(sim, schedule,
-                        initialPosition, initialSpeed, endBrakingPosition, initialSpeed * scaleFactor);
-                var speeds = new SortedDoubleMap();
-                for (var k : updatesMap.keySet()) {
-                    speeds.put(k, updatesMap.get(k).finalSpeed);
-                }
-                brakingSpeedController = new BrakingSpeedController(speeds, initialPosition, endBrakingPosition);
-            }
-            currentSpeedControllers.add(brakingSpeedController);
-            res.add(brakingSpeedController);
+        // coasting phase
+        var speed = initialSpeed;
+        var location = convertPosition(schedule, sim, initialPosition);
+        var coastingPhase = new SortedDoubleMap();
+        while (speed > targetSpeed && location.getPathPosition() < endPosition) {
+            var step = nextStep(
+                    location,
+                    speed,
+                    schedule.rollingStock,
+                    TIME_STEP,
+                    endPosition,
+                    1,
+                    (integrator) -> Action.coast());
+            coastingPhase.put(location.getPathPosition(), speed);
+            speed = step.finalSpeed;
+            location.updatePosition(schedule.rollingStock.length, step.positionDelta);
         }
-
-        // new running calculation with brakingSpeedController + scaled MapSpeedController
-        // now only the re-acceleration phase is missing
-        var newSpeeds = getExpectedSpeeds(sim, schedule, currentSpeedControllers, TIME_STEP,
-                initialPosition, endPosition, initialSpeed);
-
-        // backwards acceleration calculation
-        double speed = roiSpeeds.interpolate(endPosition); //The speed is calculated from the old running time
-        var location = convertPosition(schedule, sim, endPosition);
-        while (speed > newSpeeds.interpolate(location.getPathPosition())
-                && location.getPathPosition() - speed * TIME_STEP > 1e-5
-                && speed >= 1e-5) {
-            var directive = new SpeedDirective(newSpeeds.interpolate(location.getPathPosition()));
+        coastingPhase.put(location.getPathPosition(), targetSpeed);
+
+        // acceleration phase
+        speed = expectedSpeeds.interpolate(endPosition);
+        location = convertPosition(schedule, sim, endPosition);
+        var acceleratingPhase = new SortedDoubleMap();
+        acceleratingPhase.put(location.getPathPosition(), speed);
+        while (speed > coastingPhase.interpolate(location.getPathPosition())
+                && speed > targetSpeed && location.getPathPosition() > initialPosition) {
+            var stepSpeed = speed;
+            var directive = new SpeedDirective(coastingPhase.interpolate(location.getPathPosition()));
             var step = nextStep(
                     location,
                     speed,
```

Comment:
You compute all this, but you never read the values you put in acceleratingPhase. There's either a bug or things that can be removed

Reply:
I computed the acceleratingPhase because i need to know when it "ends" backwards, i.e. when it starts. The location of the beginning of the acceleration phase is the only result right now, but this code will need to use that acceleration phase again when we will take into account the case where the user asks for a high margin on a short distance. Then we will need to do a dichotomy on a coasting phase between a braking curve and the accelerating curve, so we need to compute and memorize them.

Reply:
You should probably explain this in a comment then

Reply:
Ok